pkg/networking/session_test.go

```diff
@@ -0,0 +1,480 @@
+	require.NoError(t, err)
+	serverSession, err := net.NewSession(ctx, serverConn, testConfig(t, mockProtocol, serverHandler, "server"))
+	require.NoError(t, err)
+
+	clientHandler.On("OnClose", clientSession).Return()
+	serverHandler.On("OnClose", serverSession).Return()
+
+	err = clientSession.Close()
+	assert.NoError(t, err)
+	err = clientSession.Close()
+	assert.NoError(t, err)
+
+	err = serverSession.Close()
+	assert.NoError(t, err)
+	err = serverSession.Close()
+	assert.NoError(t, err)
+}
+
+func TestOnClosedByOtherSide(t *testing.T) {
+	defer goleak.VerifyNone(t)
+
+	mockProtocol := netmocks.NewMockProtocol(t)
+	mockProtocol.On("EmptyHandshake").Return(&textHandshake{}, nil)
+	mockProtocol.On("IsAcceptableHandshake", &textHandshake{v: "hello"}).Once().Return(true)
+	mockProtocol.On("IsAcceptableHandshake", &textHandshake{v: "hello"}).Once().Return(true)
+	mockProtocol.On("EmptyHeader").Return(&textHeader{}, nil)
+
+	clientHandler := netmocks.NewMockHandler(t)
+	serverHandler := netmocks.NewMockHandler(t)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	clientConn, serverConn := testConnPipe()
+	net := networking.NewNetwork()
+
+	clientSession, err := net.NewSession(ctx, clientConn, testConfig(t, mockProtocol, clientHandler, "client"))
+	require.NoError(t, err)
+	serverSession, err := net.NewSession(ctx, serverConn, testConfig(t, mockProtocol, serverHandler, "server"))
+	require.NoError(t, err)
+
+	var closeWG sync.WaitGroup
+	closeWG.Add(1)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	serverHandler.On("OnClose", serverSession).Return()
+	sc1 := serverHandler.On("OnHandshake", serverSession, &textHandshake{v: "hello"}).Once().Return()
+	sc1.Run(func(_ mock.Arguments) {
+		n, wErr := serverSession.Write([]byte("hello"))
+		assert.NoError(t, wErr)
+		assert.Equal(t, 5, n)
+		go func() {
+			// Close server after client received the handshake from server.
+			closeWG.Wait() // Wait for client to receive server handshake.
+			clErr := serverSession.Close()
+			assert.NoError(t, clErr)
+			wg.Done()
+		}()
+	})
+
+	clientHandler.On("OnClose", clientSession).Return()
+
+	// Send handshake to server.
+	n, err := clientSession.Write([]byte("hello"))
+	require.NoError(t, err)
+	assert.Equal(t, 5, n)
+
+	cs1 := clientHandler.On("OnHandshake", clientSession, &textHandshake{v: "hello"}).Once().Return()
+	cs1.Run(func(_ mock.Arguments) {
+		// On receiving handshake from server, signal to close the server.
+		closeWG.Done()
+		// Try to send message to server, but it will fail because server is already closed.
+		time.Sleep(10 * time.Millisecond) // Wait for server to close.
+		_, msgErr := clientSession.Write(encodeMessage("Hello session"))
+		require.Error(t, msgErr)
+		wg.Done()
+	})
+
+	wg.Wait() // Wait for client to finish.
+	err = clientSession.Close()
+	assert.Error(t, err) // Close reports the same error, because it was registered in the send loop.
+}
+
+func TestCloseParentContext(t *testing.T) {
```

Comment:
Blinking test

<details><summary>Deadlock case</summary>
<p>

```sh
$ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=1000 -timeout=30s
ok      github.com/wavesplatform/gowaves/pkg/networking 12.561s
$ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=1000 -timeout=30s
panic: test timed out after 30s
        running tests:
                TestCloseParentContext (29s)

goroutine 569 [running]:
testing.(*M).startAlarm.func1()
        /usr/local/opt/go/libexec/src/testing/testing.go:2373 +0x385
created by time.goFunc
        /usr/local/opt/go/libexec/src/time/sleep.go:215 +0x2d

goroutine 1 [chan receive]:
testing.(*T).Run(0xc0004929c0, {0x10483c3e?, 0x1da8ba4800141b50?}, 0x1054f200)
        /usr/local/opt/go/libexec/src/testing/testing.go:1751 +0x3ab
testing.runTests.func1(0xc0004929c0)
        /usr/local/opt/go/libexec/src/testing/testing.go:2168 +0x37
testing.tRunner(0xc0004929c0, 0xc000141c70)
        /usr/local/opt/go/libexec/src/testing/testing.go:1690 +0xf4
testing.runTests(0xc000010150, {0x106bf3a0, 0x6, 0x6}, {0x10302c90?, 0x103028fa?, 0x106c5f80?})
        /usr/local/opt/go/libexec/src/testing/testing.go:2166 +0x43d
testing.(*M).Run(0xc00012b040)
        /usr/local/opt/go/libexec/src/testing/testing.go:2034 +0x64a
main.main()
        _testmain.go:57 +0x9b

goroutine 587 [select]:
github.com/wavesplatform/gowaves/pkg/networking.(*Session).sendLoop(0xc000305050)
        /home/user/gowaves/pkg/networking/session.go:212 +0xfc
github.com/wavesplatform/gowaves/pkg/execution.(*TaskGroup).Run.func1()
        /home/user/gowaves/pkg/execution/taskgroup.go:64 +0x4e
created by github.com/wavesplatform/gowaves/pkg/execution.(*TaskGroup).Run in goroutine 583
        /home/user/gowaves/pkg/execution/taskgroup.go:62 +0x87

goroutine 583 [semacquire]:
sync.runtime_Semacquire(0xc000110f20?)
        /usr/local/opt/go/libexec/src/runtime/sema.go:71 +0x25
sync.(*WaitGroup).Wait(0xc0000ca730?)
        /usr/local/opt/go/libexec/src/sync/waitgroup.go:118 +0x48
github.com/wavesplatform/gowaves/pkg/networking_test.TestCloseParentContext(0xc000492b60)
        /home/user/gowaves/pkg/networking/session_test.go:377 +0x9ea
testing.tRunner(0xc000492b60, 0x1054f200)
        /usr/local/opt/go/libexec/src/testing/testing.go:1690 +0xf4
created by testing.(*T).Run in goroutine 1
        /usr/local/opt/go/libexec/src/testing/testing.go:1743 +0x390

goroutine 541 [semacquire]:
sync.runtime_Semacquire(0x1042332e?)
        /usr/local/opt/go/libexec/src/runtime/sema.go:71 +0x25
sync.(*WaitGroup).Wait(0xc0004c4d80?)
        /usr/local/opt/go/libexec/src/sync/waitgroup.go:118 +0x48
github.com/wavesplatform/gowaves/pkg/networking_test.TestCloseParentContext.func1.1()
        /home/user/gowaves/pkg/networking/session_test.go:353 +0x29
created by github.com/wavesplatform/gowaves/pkg/networking_test.TestCloseParentContext.func1 in goroutine 586
        /home/user/gowaves/pkg/networking/session_test.go:352 +0x126

goroutine 586 [select]:
io.(*pipe).read(0xc000181c20, {0xc00034e000, 0x1000, 0x571ff3c8?})
        /usr/local/opt/go/libexec/src/io/pipe.go:57 +0xa5
io.(*PipeReader).Read(...)
        /usr/local/opt/go/libexec/src/io/pipe.go:134
github.com/wavesplatform/gowaves/pkg/networking_test.(*pipeConn).Read(0x1072ea68?, {0xc00034e000?, 0xc00030f320?, 0x2?})
        /home/user/gowaves/pkg/networking/session_test.go:401 +0x1d
bufio.(*Reader).Read(0xc000181da0, {0xc000308670, 0x4, 0x103066c9?})
        /usr/local/opt/go/libexec/src/bufio/bufio.go:241 +0x197
io.ReadAtLeast({0x10551698, 0xc000181da0}, {0xc000308670, 0x4, 0x4}, 0x4)
        /usr/local/opt/go/libexec/src/io/io.go:335 +0x90
io.ReadFull(...)
        /usr/local/opt/go/libexec/src/io/io.go:354
github.com/wavesplatform/gowaves/pkg/networking_test.(*textHeader).ReadFrom(0xc000013568, {0x10551698, 0xc000181da0})
        /home/user/gowaves/pkg/networking/session_test.go:467 +0x54
github.com/wavesplatform/gowaves/pkg/networking.(*Session).readMessage(0xc000305050, {0x572430f0, 0xc000013568})
        /home/user/gowaves/pkg/networking/session.go:305 +0x49
github.com/wavesplatform/gowaves/pkg/networking.(*Session).receive(0xc000305050)
        /home/user/gowaves/pkg/networking/session.go:273 +0x3e
github.com/wavesplatform/gowaves/pkg/networking.(*Session).receiveLoop(0xc000305050)
        /home/user/gowaves/pkg/networking/session.go:260 +0x8a
github.com/wavesplatform/gowaves/pkg/execution.(*TaskGroup).Run.func1()
        /home/user/gowaves/pkg/execution/taskgroup.go:64 +0x4e
created by github.com/wavesplatform/gowaves/pkg/execution.(*TaskGroup).Run in goroutine 583
        /home/user/gowaves/pkg/execution/taskgroup.go:62 +0x87

goroutine 585 [select]:
github.com/wavesplatform/gowaves/pkg/networking.(*Session).sendLoop(0xc000304ea0)
        /home/user/gowaves/pkg/networking/session.go:212 +0xfc
github.com/wavesplatform/gowaves/pkg/execution.(*TaskGroup).Run.func1()
        /home/user/gowaves/pkg/execution/taskgroup.go:64 +0x4e
created by github.com/wavesplatform/gowaves/pkg/execution.(*TaskGroup).Run in goroutine 583
        /home/user/gowaves/pkg/execution/taskgroup.go:62 +0x87
FAIL    github.com/wavesplatform/gowaves/pkg/networking 30.424s
FAIL

```

</p>
</details>

Reply:
It's not blinking anymore:
```bash
❯ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=10000 -timeout=5m
ok      github.com/wavesplatform/gowaves/pkg/networking 123.087s

❯ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=1000 -timeout=30s
ok      github.com/wavesplatform/gowaves/pkg/networking 12.282s

❯ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=1000 -timeout=15s
ok      github.com/wavesplatform/gowaves/pkg/networking 12.288s
```
I think fix came along the fix of TaskGroup blinking tests.

Reply:
Still encounter deadlock issue

<details><summary>Details</summary>
<p>

```sh
$ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=1000 -timeout=30s
ok      github.com/wavesplatform/gowaves/pkg/networking 12.376s
$ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=1000 -timeout=30s
ok      github.com/wavesplatform/gowaves/pkg/networking 12.286s
$ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=1000 -timeout=30s
panic: test timed out after 30s
        running tests:
                TestCloseParentContext (26s)

goroutine 2254 [running]:
testing.(*M).startAlarm.func1()
        /usr/local/opt/go/libexec/src/testing/testing.go:2373 +0x385
created by time.goFunc
        /usr/local/opt/go/libexec/src/time/sleep.go:215 +0x2d

goroutine 1 [chan receive]:
testing.(*T).Run(0xc00042a000, {0x45989cb?, 0x1b8cd758000f9b50?}, 0x4664520)
        /usr/local/opt/go/libexec/src/testing/testing.go:1751 +0x3ab
testing.runTests.func1(0xc00042a000)
        /usr/local/opt/go/libexec/src/testing/testing.go:2168 +0x37
testing.tRunner(0xc00042a000, 0xc0000f9c70)
        /usr/local/opt/go/libexec/src/testing/testing.go:1690 +0xf4
testing.runTests(0xc0000a21e0, {0x47d4420, 0x6, 0x6}, {0x4416df0?, 0x4416a5a?, 0x47db000?})
        /usr/local/opt/go/libexec/src/testing/testing.go:2166 +0x43d
testing.(*M).Run(0xc0000a9180)
        /usr/local/opt/go/libexec/src/testing/testing.go:2034 +0x64a
main.main()
        _testmain.go:57 +0x9b

goroutine 2253 [select]:
github.com/wavesplatform/gowaves/pkg/networking.(*Session).sendLoop(0xc0004f82d0)
        /home/user/gowaves/pkg/networking/session.go:220 +0xf0
github.com/wavesplatform/gowaves/pkg/execution.(*TaskGroup).Run.func1()
        /home/user/gowaves/pkg/execution/taskgroup.go:64 +0x4e
created by github.com/wavesplatform/gowaves/pkg/execution.(*TaskGroup).Run in goroutine 2249
        /home/user/gowaves/pkg/execution/taskgroup.go:62 +0x87

goroutine 2184 [semacquire]:
sync.runtime_Semacquire(0x45374ee?)
        /usr/local/opt/go/libexec/src/runtime/sema.go:71 +0x25
sync.(*WaitGroup).Wait(0xc000341400?)
        /usr/local/opt/go/libexec/src/sync/waitgroup.go:118 +0x48
github.com/wavesplatform/gowaves/pkg/networking_test.TestCloseParentContext.func1.1()
        /home/user/gowaves/pkg/networking/session_test.go:356 +0x29
created by github.com/wavesplatform/gowaves/pkg/networking_test.TestCloseParentContext.func1 in goroutine 2252
        /home/user/gowaves/pkg/networking/session_test.go:355 +0x126

goroutine 2252 [select]:
io.(*pipe).read(0xc00030c360, {0xc0004fa000, 0x1000, 0x4b315908?})
        /usr/local/opt/go/libexec/src/io/pipe.go:57 +0xa5
io.(*PipeReader).Read(...)
        /usr/local/opt/go/libexec/src/io/pipe.go:134
github.com/wavesplatform/gowaves/pkg/networking_test.(*pipeConn).Read(0x4843a68?, {0xc0004fa000?, 0x0?, 0x0?})
        /home/user/gowaves/pkg/networking/session_test.go:404 +0x1d
bufio.(*Reader).Read(0xc00030c4e0, {0xc000090b30, 0x4, 0x441a869?})
        /usr/local/opt/go/libexec/src/bufio/bufio.go:241 +0x197
io.ReadAtLeast({0x46669a8, 0xc00030c4e0}, {0xc000090b30, 0x4, 0x4}, 0x4)
        /usr/local/opt/go/libexec/src/io/io.go:335 +0x90
io.ReadFull(...)
        /usr/local/opt/go/libexec/src/io/io.go:354
github.com/wavesplatform/gowaves/pkg/networking_test.(*textHeader).ReadFrom(0xc00033ea18, {0x46669a8, 0xc00030c4e0})
        /home/user/gowaves/pkg/networking/session_test.go:470 +0x54
github.com/wavesplatform/gowaves/pkg/networking.(*Session).readMessage(0xc0004f82d0, {0x4b357fc8, 0xc00033ea18})
        /home/user/gowaves/pkg/networking/session.go:318 +0x4c
github.com/wavesplatform/gowaves/pkg/networking.(*Session).receive(0xc0004f82d0)
        /home/user/gowaves/pkg/networking/session.go:283 +0x3f
github.com/wavesplatform/gowaves/pkg/networking.(*Session).receiveLoop(0xc000431290?)
        /home/user/gowaves/pkg/networking/session.go:270 +0x46
github.com/wavesplatform/gowaves/pkg/execution.(*TaskGroup).Run.func1()
        /home/user/gowaves/pkg/execution/taskgroup.go:64 +0x4e
created by github.com/wavesplatform/gowaves/pkg/execution.(*TaskGroup).Run in goroutine 2249
        /home/user/gowaves/pkg/execution/taskgroup.go:62 +0x87

goroutine 2249 [semacquire]:
sync.runtime_Semacquire(0xc000342160?)
        /usr/local/opt/go/libexec/src/runtime/sema.go:71 +0x25
sync.(*WaitGroup).Wait(0xc0000d63c0?)
        /usr/local/opt/go/libexec/src/sync/waitgroup.go:118 +0x48
github.com/wavesplatform/gowaves/pkg/networking_test.TestCloseParentContext(0xc00042a1a0)
        /home/user/gowaves/pkg/networking/session_test.go:380 +0x9ea
testing.tRunner(0xc00042a1a0, 0x4664520)
        /usr/local/opt/go/libexec/src/testing/testing.go:1690 +0xf4
created by testing.(*T).Run in goroutine 1
        /usr/local/opt/go/libexec/src/testing/testing.go:1743 +0x390

goroutine 2251 [select]:
github.com/wavesplatform/gowaves/pkg/networking.(*Session).sendLoop(0xc0004f8120)
        /home/user/gowaves/pkg/networking/session.go:220 +0xf0
github.com/wavesplatform/gowaves/pkg/execution.(*TaskGroup).Run.func1()
        /home/user/gowaves/pkg/execution/taskgroup.go:64 +0x4e
created by github.com/wavesplatform/gowaves/pkg/execution.(*TaskGroup).Run in goroutine 2249
        /home/user/gowaves/pkg/execution/taskgroup.go:62 +0x87
FAIL    github.com/wavesplatform/gowaves/pkg/networking 30.380s
FAIL

```

</p>
</details>

Reply:
This fixes the deadlock issue.
<details><summary>Diff</summary>
<p>

```diff
diff --git a/pkg/networking/session_test.go b/pkg/networking/session_test.go
index 49759c03..6e70fc6d 100644
--- a/pkg/networking/session_test.go
+++ b/pkg/networking/session_test.go
@@ -6,6 +6,7 @@ import (
 	"errors"
 	"io"
 	"log/slog"
+	"strings"
 	"sync"
 	"testing"
 	"time"
@@ -320,10 +321,15 @@ func TestOnClosedByOtherSide(t *testing.T) {
 func TestCloseParentContext(t *testing.T) {
 	defer goleak.VerifyNone(t)
 
+	var (
+		clientToServerHello = strings.Repeat("c", textHandshakeLength)
+		serverToClientHello = strings.Repeat("s", textHandshakeLength)
+	)
+
 	mockProtocol := netmocks.NewMockProtocol(t)
 	mockProtocol.On("EmptyHandshake").Return(&textHandshake{}, nil)
-	mockProtocol.On("IsAcceptableHandshake", &textHandshake{v: "hello"}).Once().Return(true)
-	mockProtocol.On("IsAcceptableHandshake", &textHandshake{v: "hello"}).Once().Return(true)
+	mockProtocol.On("IsAcceptableHandshake", &textHandshake{v: clientToServerHello}).Once().Return(true)
+	mockProtocol.On("IsAcceptableHandshake", &textHandshake{v: serverToClientHello}).Once().Return(true)
 	mockProtocol.On("EmptyHeader").Return(&textHeader{}, nil)
 
 	clientHandler := netmocks.NewMockHandler(t)
@@ -346,12 +352,16 @@ func TestCloseParentContext(t *testing.T) {
 	var wg sync.WaitGroup
 	wg.Add(2)
 
+	clientWrittenHelloMessage := make(chan struct{})
+
 	serverHandler.On("OnClose", serverSession).Return()
-	sc1 := serverHandler.On("OnHandshake", serverSession, &textHandshake{v: "hello"}).Once().Return()
+	sc1 := serverHandler.On("OnHandshake", serverSession, &textHandshake{v: clientToServerHello}).Once().Return()
 	sc1.Run(func(_ mock.Arguments) {
-		n, wErr := serverSession.Write([]byte("hello"))
+		<-clientWrittenHelloMessage // Wait for client to send handshake
+		// Without this line client can catch session shutdown error at the moment of sending the hello message.
+		n, wErr := serverSession.Write([]byte(serverToClientHello))
 		assert.NoError(t, wErr)
-		assert.Equal(t, 5, n)
+		assert.Equal(t, textHandshakeLength, n)
 		go func() {
 			closeWG.Wait() // Wait for client to receive server handshake.
 			cancel()       // Close parent context.
@@ -360,13 +370,7 @@ func TestCloseParentContext(t *testing.T) {
 	})
 
 	clientHandler.On("OnClose", clientSession).Return()
-
-	// Send handshake to server.
-	n, err := clientSession.Write([]byte("hello"))
-	require.NoError(t, err)
-	assert.Equal(t, 5, n)
-
-	cs1 := clientHandler.On("OnHandshake", clientSession, &textHandshake{v: "hello"}).Once().Return()
+	cs1 := clientHandler.On("OnHandshake", clientSession, &textHandshake{v: serverToClientHello}).Once().Return()
 	cs1.Run(func(_ mock.Arguments) {
 		// On receiving handshake from server, signal to close the server.
 		closeWG.Done()
@@ -377,6 +381,12 @@ func TestCloseParentContext(t *testing.T) {
 		wg.Done()
 	})
 
+	// Send handshake to server.
+	n, err := clientSession.Write([]byte(clientToServerHello))
+	require.NoError(t, err)
+	assert.Equal(t, textHandshakeLength, n)
+	close(clientWrittenHelloMessage) // Signal that handshake was sent, server can proceed.
+
 	wg.Wait() // Wait for client to finish.
 
 	err = clientSession.Close()
@@ -431,13 +441,15 @@ func encodeMessage(s string) []byte {
 	return msg
 }
 
+const textHandshakeLength = 5
+
 // We have to use the "real" handshake, not a mock, because we are reading or writing to a "real" piped connection.
 type textHandshake struct {
 	v string
 }
 
 func (h *textHandshake) ReadFrom(r io.Reader) (int64, error) {
-	buf := make([]byte, 5)
+	buf := make([]byte, textHandshakeLength)
 	n, err := io.ReadFull(r, buf)
 	if err != nil {
 		return int64(n), err

```

</p>
</details> 

```sh
$ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=1000 -timeout=30s
ok      github.com/wavesplatform/gowaves/pkg/networking 12.311s
$ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=1000 -timeout=30s
ok      github.com/wavesplatform/gowaves/pkg/networking 12.373s
$ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=10000 -timeout=150s
ok      github.com/wavesplatform/gowaves/pkg/networking 119.805s
$ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=10000 -timeout=150s
ok      github.com/wavesplatform/gowaves/pkg/networking 119.607s
```

Reply:
Fixed. I implemented the essence of your fix. Actually sometimes the parent context was closed before the client 
finished to send handshake. Sounds strange but it was a chain of nested handlers like: [Client Send Handshake] -> [Server Receive Handshake] -> [Server Send Handshake]. And in the inner one the context was cancelled that leads to not finishing the outermost one. So, breaking the chain by introducing the new wait group (channel in your case) was the solution.
<details><summary>Test runs:</summary>
<p>

```bash
❯ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=1000 -timeout=30s
ok      github.com/wavesplatform/gowaves/pkg/networking 12.676s

❯ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=1000 -timeout=30s
ok      github.com/wavesplatform/gowaves/pkg/networking 12.510s

❯ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=1000 -timeout=30s
ok      github.com/wavesplatform/gowaves/pkg/networking 12.693s

❯ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=1000 -timeout=30s
ok      github.com/wavesplatform/gowaves/pkg/networking 12.352s

❯ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=1000 -timeout=30s
ok      github.com/wavesplatform/gowaves/pkg/networking 12.630s

❯ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=10000 -timeout=300s
ok      github.com/wavesplatform/gowaves/pkg/networking 125.723s

❯ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=1000 -timeout=30s -race
ok      github.com/wavesplatform/gowaves/pkg/networking 15.957s

❯ go test github.com/wavesplatform/gowaves/pkg/networking -run "TestCloseParentContext" -failfast -count=10000 -timeout=300s -race
ok      github.com/wavesplatform/gowaves/pkg/networking 157.216s
```

</p>
</details>